File: ElegantAppluanch/src/main/java/com/jacob/material/applaunch/MainActivity.java
package com.jacob.material.applaunch;

import android.os.Bundle;
import android.util.Log;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.jacob.http.HttpLoader;
import com.jacob.http.MdRetrofitFactory;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class MainActivity extends AppCompatActivity {

    private ThronesAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.setContentView(R.layout.main_activity);

        RecyclerView recyclerView = findViewById(R.id.recycler_view);

        LinearLayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
        recyclerView.setLayoutManager(layoutManager);
        adapter = new ThronesAdapter(this);
        recyclerView.setAdapter(adapter);

        loadThrones();
    }

    private void loadThrones(){

        MdRetrofitFactory.getLoader(HttpLoader.class).getThrones()
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new io.reactivex.Observer<List<Thrones>>(){
                    @Override
                    public void onSubscribe(Disposable d) {
                        Log.d("MATERIAL_DESIGN", "onSubscribe");
                    }

                    @Override
                    public void onNext(List<Thrones> list) {
                        Log.d("MATERIAL_DESIGN", "onNext list size = " + list.size());
                        adapter.setDataList(list);
                    }

                    @Override
                    public void onError(Throwable e) {
                        Log.d("MATERIAL_DESIGN", "onError");
                        e.printStackTrace();
                    }

                    @Override
                    public void onComplete() {
                        Log.d("MATERIAL_DESIGN", "onComplete");
                    }
                });
    }

}

File: TerribleAppLaunch/src/main/java/com/jacob/material/terriblelaunch/MainActivity.kt
package com.jacob.material.terriblelaunch

import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.jacob.http.HttpLoader
import com.jacob.http.MdRetrofitFactory.getLoader
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import retrofit2.Response

class MainActivity : AppCompatActivity() {
    private var adapter: ThronesAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        this.setContentView(R.layout.main_activity)

        val recyclerView = findViewById<RecyclerView>(R.id.recycler_view)
        val layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        recyclerView.layoutManager = layoutManager
        adapter = ThronesAdapter(this)
        recyclerView.adapter = adapter
        loadThrones()
    }

    private fun loadThrones() {
        lifecycleScope.launch {
            var response:Response<List<Thrones>>
            withContext(Dispatchers.IO){
                response = getLoader(HttpLoader::class.java).getThrones();
            }
            if(response.isSuccessful){
                adapter!!.setDataList(response.body());

            }
        }
    }
}